python/lsst/pipe/base/pipeline.py

```diff
@@ -680,7 +709,55 @@ def __len__(self):
     def __eq__(self, other: object):
         if not isinstance(other, Pipeline):
             return False
-        return self._pipelineIR == other._pipelineIR
+        elif self._pipelineIR == other._pipelineIR:
+            # Shortcut: if the IR is the same, the expanded pipeline must be
+            # the same as well.  But the converse is not true.
+            return True
+        else:
+            self_expanded = {td.label: (td.taskClass,) for td in self}
+            other_expanded = {td.label: (td.taskClass,) for td in other}
+            if self_expanded != other_expanded:
+                return False
+        # After DM-27847, we should compare configuration here, or better,
+        # delegated to TaskDef.__eq__ after making that compare configurations.
+        raise NotImplementedError(
+            "Pipelines cannot be compared because config instances cannot be compared; see DM-27847."
+        )
+
+    def _write_expanded_dir(self, uri: ButlerURI, task_defs: Optional[Iterable[TaskDef]] = None) -> None:
+        """Internal implementation of `write_to_uri` with ``expand=True`` and
+        a directory-like URI.
+
+        Parameters
+        ----------
+        uri : `str` or `ButlerURI`
+            URI to write to; may have any scheme with `ButlerURI` write or no
+            scheme for a local file/directory.  Should have a trailing slash
+            (indicating a directory-like URI).
+        task_defs : `Iterable` [ `TaskDef` ], optional
+            Output of `toExpandedPipeline`; may be passed to avoid a second
+            call to that method internally.
+        """
+        assert uri.dirLike, f"{uri} is not a directory-like URI."
+        # Expand the pipeline.  This applies all config overrides, applies all
+        # parameters, checks contracts, and sorts tasks topologically with
+        # lexicographical (on label) tiebreaking.
+        if task_defs is not None:
+            task_defs = list(task_defs)
+        else:
+            task_defs = list(self.toExpandedPipeline())
+        uri.mkdir()
+        config_dir_uri = uri.join("config/")
+        config_dir_uri.mkdir()
+        expanded_tasks: Dict[str, pipelineIR.TaskIR] = {}
+        for task_def in task_defs:
+            task_ir = pipelineIR.TaskIR(label=task_def.label, klass=task_def.taskName, config=[])
+            config_uri = config_dir_uri.join(f"{task_def.label}.py")
+            with config_uri.open("w") as buffer:
+                task_def.config.saveToStream(buffer)
+            task_ir.config.append(pipelineIR.ConfigIR(file=[f"config/{task_def.label}.py"]))
```

Comment:
Just thought it was easier to read the logic if I put it after I actually wrote the config file, but since you found it surprising instead, I'll change it.